devito/arguments.py

```diff
@@ -158,198 +143,396 @@ class PtrArgument(Argument):
+        for i in self.dle_arguments:
+            dim_size = dim_sizes.get(i.original_dim.name, None)
+            if dim_size is None:
+                raise InvalidArgument('Unable to derive size of dimension %s from '
+                                      'defaults. Please provide an explicit '
+                                      'value.' % i.original_dim.name)
+            if i.value:
+                try:
+                    dle_arguments[i.argument.name] = i.value(dim_size)
+                except TypeError:
+                    dle_arguments[i.argument.name] = i.value
+                    autotune = False
+            else:
+                dle_arguments[i.argument.name] = dim_size
+        return dle_arguments, autotune
+
+    def _derive_values(self, kwargs):
+        """ Populate values for all the arguments. The values provided in kwargs will
+            be used wherever provided. The values for the rest of the arguments will be
+            derived from the ones provided. The default values for the tensors come from
+            the data property of the symbols used in the Operator.
         """
-        raise NotImplemented()
+        values = OrderedDict()
+
+        values = OrderedDict([(i, get_value(i, kwargs.pop(i.name, None), values))
+                              for i in self.arguments])
+
+        dimension_values = OrderedDict([(i, kwargs.pop(i.name, None))
+                                        for i in self.dimension_params])
+
+        # Make sure we've used all arguments passed
+        if len(kwargs) > 0:
+            raise InvalidArgument("Unknown arguments passed: " + ", ".join(kwargs.keys()))
+
+        # Derive values for other arguments
+        for i in self.arguments:
+            if values[i] is None:
+                known_values = OrderedDict(chain(values.items(),
+                                                 dimension_values.items()))
+                provided_values = [get_value(i, x, known_values)
+                                   for x in i.gets_value_from]
+                assert(len(provided_values) > 0)
+                values[i] = reduce(max, provided_values)
+
+        # Second pass to evaluate any Unevaluated dependencies from the first pass
+        for k, v in values.items():
+            if isinstance(v, UnevaluatedDependency):
+                values[k] = v.evaluate(values)
+        return values
+
+    def _verify(self, values):
+        verify = True
+        for i in values:
+            verify = verify and all(verify(i, x) for x in i.verified_by)
+        return verify
 
+    @property
+    def dimensions(self):
+        return [x for x in self.parameters if isinstance(x, Dimension)]
 
-class DimensionArgProvider(ArgumentProvider):
 
-    """ This class is used to decorate the Dimension class with behaviour required
-        to handle runtime arguments. All properties/methods defined here are available
-        in any Dimension object.
+class ArgumentVisitor(GenericVisitor):
```

Comment:
I actually got one final question: why is this a tree visitor if you never call ``self.visit(child/children)`` ?

Reply:
So it is not a tree visitor, I guess. I'll change the docstring on the parent class. Is that all?

Reply:
> So it is not a tree visitor, I guess. I'll change the docstring on the parent class. Is that all?

Is this equivalent to:
```
for i
  if i is ... do()
  elif i is ... do()
  elif i is ... do()
  else: 
```
 If there is no concept of visit (ie, there is never a tree, or the input is always leaf), then using a Visitor here is really misleading. At least, that's the way I conceived those Visitors.

Reply:
The code you suggest is dirtier than the clean separation offered in Visitors. Visitors are not only used for trees but also to encapsulate an operation being performed differently on classes of different types. From the [python docs](http://python-3-patterns-idioms-test.readthedocs.io/en/latest/Visitor.html): 
> The visitor pattern allows you to extend the interface of the primary type by creating a separate class hierarchy of type Visitor to virtualize the operations performed upon the primary type.

Reply:
I was **not** suggesting, I was only asking to understand.